mermithid/processors/misc/ComplexLineShape.py

```diff
@@ -0,0 +1,668 @@
+            current_full_spectrum += current_working_spectrum*(prob_parameter*p)**(n)
+
+            current_working_spectrum = \
+            scatter_spectra.item()['{}_{}'.format(gases[0], gases[1])]\
+            ['{}_{}'.format(str(0).zfill(2), str(n).zfill(2))]
+            current_working_spectrum = \
+            self.normalize(signal.convolve(zeroth_order_peak, current_working_spectrum, mode='same'))
+            current_full_spectrum += current_working_spectrum*(prob_parameter*q)**(n)
+        return current_full_spectrum
+
+    # Produces a spectrum in real energy that can now be evaluated off of the SELA.
+    #def spectrum_func(x_keV,FWHM_G_eV,line_pos_keV,scatter_prob,amplitude):
+    def spectrum_func(self, x_keV, *p0):
+        x_eV = x_keV*1000.
+        en_loss_array = self.std_eV_array()
+        en_loss_array_min = en_loss_array[0]
+        en_loss_array_max = en_loss_array[len(en_loss_array)-1]
+        en_array_rev = ComplexLineShapeUtilities.flip_array(-1*en_loss_array)
+        f = np.zeros(len(x_keV))
+        f_intermediate = np.zeros(len(x_keV))
+
+        FWHM_G_eV = p0[0]
+        line_pos_keV = p0[1]
+        amplitude = p0[2]
+        prob_parameter = p0[3]
+        scatter_proportion = p0[4]
+    
+        line_pos_eV = line_pos_keV*1000.
+        x_eV_minus_line = x_eV - line_pos_eV
+        zero_idx = np.r_[np.where(x_eV_minus_line< en_loss_array_min)[0],np.where(x_eV_minus_line>en_loss_array_max)[0]]
+        nonzero_idx = [i for i in range(len(x_keV)) if i not in zero_idx]
+    
+        full_spectrum = self.make_spectrum(FWHM_G_eV, prob_parameter, scatter_proportion)
+        full_spectrum_rev = ComplexLineShapeUtilities.flip_array(full_spectrum)
+        f_intermediate[nonzero_idx] = np.interp(x_eV_minus_line[nonzero_idx],en_array_rev,full_spectrum_rev)
+        f[nonzero_idx] += amplitude*f_intermediate[nonzero_idx]/np.sum(f_intermediate[nonzero_idx])
+        return f
+
+    # Call this function to fit a histogram of start frequencies with the model.
+    # Note that the data_hist_freq should be the StartFrequencies as given by katydid,
+    # which will be from ~0 MHZ to ~100 MHz. You must also pass this function the
+    # self.RF_ROI_MIN value from the metadata file of your data.
+    # You must also supply a guess for the self.B_field present for the run;
+    # 0.959 T is usually sufficient.
+
+    def fit_data(self, freq_bins, data_hist_freq, print_params=True):
+        t = time.time()
+        self.check_existence_of_scatter_file()
+        bins_Hz = freq_bins + self.RF_ROI_MIN
+        bins_keV = ComplexLineShapeUtilities.frequency_to_energy(bins_Hz,self.B_field)
+        bins_keV = ComplexLineShapeUtilities.flip_array(bins_keV)
+        data_hist = ComplexLineShapeUtilities.flip_array(data_hist_freq)
+        bins_keV_nonzero , data_hist_nonzero , data_hist_err = ComplexLineShapeUtilities.get_only_nonzero_bins(bins_keV, data_hist)
```

Comment:
You shouldn't remove empty data bins for your analysis: it could be that the number of predicted events in a bin is small, but non-zero and so you measured zero events in a bin.
I think removing empty bins will bias the result of your analysis.
I guess you removed these to avoid having to deal with zero error in the chi2 computation: instead you should compute the error of the model or use a binned Likelihood technique as discussed in another comment...